### PR TITLE
ENGINE-2472: Allows creation of custom output parameters based on columns list

### PR DESCRIPTION
**Pull request by aklotos**

### CODE REVIEW

File: thng-resource-model/src/main/java/com/evrythng/thng/resource/model/store/BatchPopulatingTask.java [hunk 1]
```diff
@@ -288,9 +291,12 @@ public void setOutputParameters(final OutputParameters outputParameters) {
 
 		void setType(final Type type);
 
+		/**
+		 * Type of supported {@link BatchPopulatingTask} input parameters
+		 */
 		enum Type {
 
-			FIXED_AMOUNT, FILE_BASED, PRIVATE_FILE_BASED, LIST_BASED
+			FIXED_AMOUNT, FILE_BASED, PRIVATE_FILE_BASED, LIST_BASED, IDENTIFIERS_LIST_BASED
```
Comment (tatacalu) [inline thng-resource-model/src/main/java/com/evrythng/thng/resource/model/store/BatchPopulatingTask.java]: Javadoc please

File: thng-resource-model/src/main/java/com/evrythng/thng/resource/model/store/IdentifiersListBasedAdiInputParameters.java [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+/**
+ * Represents {@link BatchPopulatingTask} input parameters based on list of identifiers
+ */
+public class IdentifiersListBasedAdiInputParameters extends BaseAdiInputParameters {
+
+	public static final String FIELD_CUSTOM_IDENTIFIER = "customIdentifier";
+	private IdentifiersResource customIdentifier;
+
+	public IdentifiersResource getCustomIdentifier() {
```
Comment (tatacalu) [inline thng-resource-model/src/main/java/com/evrythng/thng/resource/model/store/IdentifiersListBasedAdiInputParameters.java]: Javadoc here and on the getter as well.

File: thng-resource-model/src/main/java/com/evrythng/thng/resource/model/store/IdentifiersResource.java [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+/**
+ * Represents list of identifiers resource with the same key.
+ */
+public class IdentifiersResource {
+
+	public static final String FIELD_KEY = "key";
+	public static final String FIELD_VALUES = "values";
+
+	private String key;
+	private java.util.Collection<String> values;
```
Comment (tatacalu) [inline thng-resource-model/src/main/java/com/evrythng/thng/resource/model/store/IdentifiersResource.java]: please import java.util.Collection

File: thng-resource-model/src/main/java/com/evrythng/thng/resource/model/store/IdentifiersResource.java [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+	public IdentifiersResource() {
```
Comment (tatacalu) [inline thng-resource-model/src/main/java/com/evrythng/thng/resource/model/store/IdentifiersResource.java]: do we still need this public if we have of() ?

File: thng-resource-model/src/main/java/com/evrythng/thng/resource/model/store/IdentifiersResource.java [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+	public java.util.Collection<String> getValues() {
+
+		return values;
+	}
+
+	public void setValues(final java.util.Collection<String> values) {
+
+		this.values = values;
+	}
+}
```
Comment (tatacalu) [inline thng-resource-model/src/main/java/com/evrythng/thng/resource/model/store/IdentifiersResource.java]: A new line is needed

File: thng-resource-model/src/main/java/com/evrythng/thng/resource/model/store/BatchPopulatingTask.java (hunk 1 above)
Comment (aklotos) [inline thng-resource-model/src/main/java/com/evrythng/thng/resource/model/store/BatchPopulatingTask.java]: there is a javadoc on Enum, you want javadoc on every enum value, right?

Comment (tatacalu) [inline thng-resource-model/src/main/java/com/evrythng/thng/resource/model/store/BatchPopulatingTask.java]: yes

Comment (aklotos) [inline thng-resource-model/src/main/java/com/evrythng/thng/resource/model/store/BatchPopulatingTask.java]: we definitely need a cathup with other java guys to agree on wiriting javadocs, what should go with and what can go without it

Comment (tatacalu) [inline thng-resource-model/src/main/java/com/evrythng/thng/resource/model/store/BatchPopulatingTask.java]: do you think we should not document Enum values ?

File: thng-resource-model/src/main/java/com/evrythng/thng/resource/model/store/IdentifiersResource.java [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+import java.util.Collection;
+
+/**
+ * Represents list of identifiers resource with the same key.
+ */
+public class IdentifiersResource {
```
Comment (dorogush) [inline thng-resource-model/src/main/java/com/evrythng/thng/resource/model/store/IdentifiersResource.java]: It should be either mutable pojo (no args constructor) or immutable value class (final fields). And anyway `equal` and `hashCode` should be implemented. `toString` is good to be added as well.

Comment (aklotos) [inline thng-resource-model/src/main/java/com/evrythng/thng/resource/model/store/IdentifiersResource.java]: I don't like the idea discussing something in already closed and merged PRs. Agree with mutable/immutable point. Disagree with `equals` and `hashCode`. This class is not designed to check equality with other object, it's also not designed to be used as a key for `Set` or `Map`. We don't use this class this way, we don't want other developers to use this class this way.

Comment (tatacalu) [inline thng-resource-model/src/main/java/com/evrythng/thng/resource/model/store/IdentifiersResource.java]: I would have to agree with Sasha at this point regarding equals / hashCode and toString as well. It is a public class in the public Java SDK, users are going to use it however they want, we have no control over them unfortunately.

Comment (aklotos) [inline thng-resource-model/src/main/java/com/evrythng/thng/resource/model/store/IdentifiersResource.java]: So we want to put the gun into their hands and aim it to their legs? It is public, right, but I would prefer leave a javadoc note that this class is not intended to be used for equality check and as a key of other collections and therefore it doesn't have implemented these methods.

**Comment by tatacalu** [comment]
As I said, anyone can use this class however they like and because of that we have to limit the potential misbehaviour. A javadoc comment will not prevent them from that. In any case, equals / hashCode / toString methods are standard things, I fail to understand why you are so against them?

**Comment by aklotos** [comment]
I just consider this incorrect. I used `Collection` as a property of resource, it's not intended to check the equality. Let's define what is equal for `Collection` of elements. The same links? The same elements? How do you check this? `O(n^2)` loop? What if this is not Collection and List? Does the order of elements matter? What if user wants another way?
Probably I'm not right but in my opinion we should just say that if you want to use this resource this way you should define your own way to do `equals` or `hashCode`.

**Comment by aklotos** [comment]
Anyway if considering all my arguments you both still have your point I'll obey and add `equals` and `hashCode`.

**Comment by coffeemachine** [comment]
As a general rule of thumb, value objects are considered equal if their fields are equal. This looks like a value object to me. However, the behavior should be consistent across the all of the wrapper classes.

In terms of implementation, you may want to consider if the order of the elements is important or not, if duplicated values make sense, and eventually if the field type should be Collection at all.

**Comment by aklotos** [comment]
Now I obey. Thanks for discussion!

**Pull request merged**